File: ios/GradeThread/Grading/CertIntegrity.swift
import Foundation

/// US-1294: certificate-integrity verification — the iOS twin of the web
/// `/api/content/public/certificates/:id/verify` flow + the certificate page's
/// IntegrityPanel. Re-derives nothing client-side; it asks the public edge
/// endpoint (which recomputes the sealed content hash + checks the HMAC) and
/// renders the verdict so a seller/buyer can confirm a Digital Slab (US-763) is
/// authentic before trusting — or sharing — it.
///
/// Everything here is PURE (no networking) so it unit-tests with no env; the
/// network call lives in ``CertIntegrityService``.

/// Raw verify-endpoint response. Snake-case keys are mapped by the EdgeAPI
/// decoder (`.convertFromSnakeCase`). `issuedAt` stays a `String`: the edge
/// emits fractional-second ISO timestamps that the default `.iso8601` date
/// strategy REJECTS, and we only ever display it.
struct CertVerifyResponse: Decodable, Equatable {
    let certificateId: String
    /// "verified" | "mismatch" | "unverifiable" (matches edge `verifyCertIntegrity`).
    let status: String
    let verified: Bool
    let signed: Bool
    let algorithm: String
    let integrityVersion: Int?
    let contentHash: String?
    let issuedAt: String?
}

/// The resolved verdict the UI renders. Collapses the endpoint status PLUS any
/// transport/parse failure into one exhaustive set so a tampered OR unreachable
/// certificate can NEVER be presented as a silent pass (AC3).
enum CertVerification: Equatable {
    /// In-flight — the verify request hasn't resolved yet.
    case verifying
    /// Hash (and signature, when signed) matched the sealed record.
    case verified(signed: Bool)
    /// status "mismatch": grade claims don't hash to the sealed value (tamper/forgery).
    case tampered
    /// status "unverifiable": legacy grade finalized before the integrity scheme.
    case unverifiable
    /// Couldn't reach or parse the verify endpoint (offline, 404/withheld, 5xx).
    case unavailable
}

/// Pure helpers for the verification flow.
enum CertIntegrity {
    /// Maps the endpoint response → the rendered verdict. FAIL CLOSED: only an
    /// explicit `status == "verified"` AND `verified == true` is a pass; an
    /// explicit "mismatch" is tampered; everything else (including a stray
    /// `verified:false`) degrades to `.unverifiable` so nothing reads as authentic.
    static func classify(_ response: CertVerifyResponse) -> CertVerification {
        if response.status == "verified", response.verified {
            return .verified(signed: response.signed)
        }
        if response.status == "mismatch" {
            return .tampered
        }
        return .unverifiable
    }
}

/// Pulls the `<id>` out of a public certificate URL (`<site>/cert/<id>`,
/// optionally with a query/fragment). Returns nil for a non-certificate URL —
/// the caller then renders `.unavailable` rather than hitting a bogus path.
enum CertificateID {
    static func extract(fromCertificateURL url: String?) -> String? {
        guard let raw = url?.trimmingCharacters(in: .whitespacesAndNewlines),
              !raw.isEmpty,
              let range = raw.range(of: "/cert/") else { return nil }
        var rest = String(raw[range.upperBound...])
        // Stop at the first query/fragment/path delimiter — a UUID never
        // contains one, so the id is whatever precedes it.
        if let end = rest.firstIndex(where: { $0 == "?" || $0 == "#" || $0 == "/" }) {
            rest = String(rest[..<end])
        }
        return rest.isEmpty ? nil : rest
    }
}

/// Render-ready presentation for a verdict (copy + glyph + tone). Pure so the
/// badge view stays declarative and the copy is unit-testable.
struct CertVerificationDisplay: Equatable {
    enum Tone: Equatable { case verified, danger, warning, neutral }

    let title: String
    let detail: String
    let systemImage: String
    let tone: Tone
    /// True when offering a "Try again" affordance makes sense (transient failure).
    let retryable: Bool
}

extension CertVerification {
    var display: CertVerificationDisplay {
        switch self {
        case .verifying:
            return CertVerificationDisplay(
                title: "Verifying certificate…",
                detail: "Confirming this grade matches its sealed record.",
                systemImage: "shield.lefthalf.filled",
                tone: .neutral,
                retryable: false)
        case let .verified(signed):
            return CertVerificationDisplay(
                title: "Certificate verified",
                detail: signed
                    ? "Grade claims match the cryptographically signed record."
                    : "Grade claims match the sealed record.",
                systemImage: "checkmark.seal.fill",
                tone: .verified,
                retryable: false)
        case .tampered:
            return CertVerificationDisplay(
                title: "Integrity check failed",
                detail: "These grade claims don't match the sealed record — don't trust this certificate.",
                systemImage: "exclamationmark.octagon.fill",
                tone: .danger,
                retryable: false)
        case .unverifiable:
            return CertVerificationDisplay(
                title: "Integrity record unavailable",
                detail: "This grade predates tamper-proofing, so it can't be cryptographically verified.",
                systemImage: "questionmark.shield.fill",
                tone: .warning,
                retryable: false)
        case .unavailable:
            return CertVerificationDisplay(
                title: "Couldn't verify",
                detail: "We couldn't reach the verification service. Check your connection and try again.",
                systemImage: "wifi.exclamationmark",
                tone: .warning,
                retryable: true)
        }
    }

    /// True only for a confirmed pass — callers must treat every other case as
    /// a non-verified state (AC3: never a silent pass).
    var isVerified: Bool {
        if case .verified = self { return true }
        return false
    }
}

File: android/app/src/main/java/com/gradethread/app/grading/CertIntegrity.kt
package com.gradethread.app.grading

import androidx.annotation.StringRes
import com.gradethread.app.R

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * US-1337: certificate-integrity verification (iOS `CertIntegrity`, US-1294).
 *
 * Nothing is re-derived on the device. The public edge endpoint recomputes the
 * sealed content hash and checks the HMAC; this classifies its verdict. Pure,
 * so the fail-closed rule is provable with no network.
 */
@Serializable
data class CertVerifyResponse(
    @SerialName("certificate_id") val certificateId: String = "",
    /** "verified" | "mismatch" | "unverifiable" — the edge's own vocabulary. */
    val status: String = "",
    val verified: Boolean = false,
    val signed: Boolean = false,
    val algorithm: String = "",
    @SerialName("integrity_version") val integrityVersion: Int? = null,
    @SerialName("content_hash") val contentHash: String? = null,
    /**
     * Left as a String on purpose: the edge emits fractional-second ISO
     * timestamps that strict parsers reject, and this is only ever displayed.
     */
    @SerialName("issued_at") val issuedAt: String? = null,
)

/**
 * The rendered verdict. Transport failure collapses into the same set as the
 * endpoint's own statuses so a tampered OR unreachable certificate can never
 * be presented as a silent pass.
 */
sealed class CertVerification {
    object Verifying : CertVerification()
    data class Verified(val signed: Boolean) : CertVerification()

    /** The grade claims don't hash to the sealed value — tamper or forgery. */
    object Tampered : CertVerification()

    /** Legacy grade, finalized before the integrity scheme existed. */
    object Unverifiable : CertVerification()

    /** Offline, 404 (withheld / non-public), or 5xx. */
    object Unavailable : CertVerification()

    /** True ONLY for a confirmed pass. Everything else is a non-verified state. */
    val isVerified: Boolean get() = this is Verified
}

object CertIntegrity {

    /**
     * Map a response to a verdict. FAILS CLOSED.
     *
     * Only an explicit `status == "verified"` AND `verified == true` passes.
     * An explicit "mismatch" is tamper; everything else — including a response
     * that says "verified" while the boolean is false — degrades to
     * [CertVerification.Unverifiable]. The asymmetry is deliberate: the cost of
     * calling a forgery authentic is far higher than the cost of declining to
     * vouch for a genuine certificate.
     */
    fun classify(response: CertVerifyResponse): CertVerification = when {
        response.status == "verified" && response.verified ->
            CertVerification.Verified(signed = response.signed)

        response.status == "mismatch" -> CertVerification.Tampered

        else -> CertVerification.Unverifiable
    }

    /**
     * Render-ready copy for a verdict.
     *
     * US-2976: string RESOURCES. This banner is the one place the app says
     * whether a certificate can be TRUSTED, so a Spanish buyer reading
     * "Integrity check failed" in English is being shown the most important
     * sentence on the screen in a language they may not have.
     */
    fun display(verification: CertVerification): Display = when (verification) {
        CertVerification.Verifying -> Display(
            title = R.string.cert_integrity_verifying,
            detail = R.string.cert_integrity_verifying_detail,
            tone = Tone.NEUTRAL,
        )

        is CertVerification.Verified -> Display(
            title = R.string.cert_integrity_verified,
            detail = if (verification.signed) {
                R.string.cert_integrity_verified_signed_detail
            } else {
                R.string.cert_integrity_verified_detail
            },
            tone = Tone.VERIFIED,
        )

        CertVerification.Tampered -> Display(
            title = R.string.cert_integrity_tampered,
            detail = R.string.cert_integrity_tampered_detail,
            tone = Tone.DANGER,
        )

        CertVerification.Unverifiable -> Display(
            title = R.string.cert_integrity_unverifiable,
            detail = R.string.cert_integrity_unverifiable_detail,
            tone = Tone.WARNING,
        )

        CertVerification.Unavailable -> Display(
            title = R.string.cert_integrity_unavailable,
            detail = R.string.cert_integrity_unavailable_detail,
            tone = Tone.WARNING,
            retryable = true,
        )
    }

    enum class Tone { VERIFIED, DANGER, WARNING, NEUTRAL }

    data class Display(
        @StringRes val title: Int,
        @StringRes val detail: Int,
        val tone: Tone,
        val retryable: Boolean = false,
    )
}

/** Pulls the id out of a public certificate URL (`<site>/cert/<id>`). */
object CertificateId {

    fun extract(certificateUrl: String?): String? {
        val raw = certificateUrl?.trim().orEmpty()
        if (raw.isEmpty()) return null
        val marker = raw.indexOf(MARKER)
        if (marker < 0) return null
        var rest = raw.substring(marker + MARKER.length)
        // Stop at the first delimiter — a UUID never contains one, so the id is
        // whatever precedes it.
        val end = rest.indexOfFirst { it == '?' || it == '#' || it == '/' }
        if (end >= 0) rest = rest.substring(0, end)
        return rest.ifEmpty { null }
    }

    private const val MARKER = "/cert/"
}
